BLASTn was used to identify homologous sequences for each Nanopore amplicon consensus against the NCBI non-redundant database (Altschul 1997; NCBI 1988). The top hits (lowest e-value) were downloaded and aligned with the amplicon consensus sequences using MAFFT v7 (Katoh and Standley 2013), then trimmed with BMGE -g 0.3 (Criscuolo and Gribaldo 2010). Phylogenies for the different recovered taxa were constructed separately for SSU and LSU genes, when possible. Maximum likelihood trees were built with FastTree v2.1.11 and the GTR model (gamma option, 1000 bootstraps) (Price, Dehal, and Arkin 2009; Abadi et al. 2019). Trees were visualised with the ape and ggtree R packages and eventually edited in Adobe Illustrator (Paradis and Schliep 2019; Yu et al. 2017; R Core Team 2022).
One concatenated tree was constructed by creating a matrix from gregarine consensuses recovered from the same host that had both 18S and 28S data. For this, trimmed and curated sequences were concatenated with FASconCAT-G (Kück and Longo 2014) and the tree was made with the same method as above. For the gregarine host co-phylogeny, one representative nucleotide sequence of the cytochrome c oxidase subunit I (COI) gene for each host species was downloaded from GenBank. A maximum likelihood tree with 1,000 bootstraps was constructed as above using MAFFT and BMGE, with the exception of using iqtree2 and the GTR+G5+I model (Pentinsaari et al. 2016; Goh et al. 2022; Hendrich et al. 2010; Minh et al. 2020).

### References:
Abadi, Shiran, Dana Azouri, Tal Pupko, and Itay Mayrose. 2019. ‘Model Selection May Not Be a Mandatory Step for Phylogeny Reconstruction’. Nature Communications 10 (1): 934. https://doi.org/10.1038/s41467-019-08822-w.
Altschul, S. 1997. ‘Gapped BLAST and PSI-BLAST: A New Generation of Protein Database Search Programs’. Nucleic Acids Research 25 (17): 3389–3402. https://doi.org/10.1093/nar/25.17.3389.
Criscuolo, Alexis, and Simonetta Gribaldo. 2010. ‘BMGE (Block Mapping and Gathering with Entropy): A New Software for Selection of Phylogenetic Informative Regions from Multiple Sequence Alignments’. BMC Evolutionary Biology 10 (1): 210. https://doi.org/10.1186/1471-2148-10-210.
Goh, King-Siang, Liang-Jong Wang, Jing-Han Ni, and Tzi-Yuan Wang. 2022. ‘Luminescent Characteristics and Mitochondrial COI Barcodes of Nine Cohabitated Taiwanese Fireflies’. PeerJ 10 (October):e14195. https://doi.org/10.7717/peerj.14195.
Hendrich, Lars, Joan Pons, Ignacio Ribera, and Michael Balke. 2010. ‘Mitochondrial Cox1 Sequence Data Reliably Uncover Patterns of Insect Diversity But Suffer from High Lineage-Idiosyncratic Error Rates’. Edited by Carles Lalueza-Fox. PLoS ONE 5 (12): e14448. https://doi.org/10.1371/journal.pone.0014448.
Katoh, K., and D. M. Standley. 2013. ‘MAFFT Multiple Sequence Alignment Software Version 7: Improvements in Performance and Usability’. Molecular Biology and Evolution 30 (4): 772–80. https://doi.org/10.1093/molbev/mst010.
Kück, Patrick, and Gary C Longo. 2014. ‘FASconCAT-G: Extensive Functions for Multiple Sequence Alignment Preparations Concerning Phylogenetic Studies’. Frontiers in Zoology 11 (1): 81. https://doi.org/10.1186/s12983-014-0081-x.
Minh, Bui Quang, Heiko A Schmidt, Olga Chernomor, Dominik Schrempf, Michael D Woodhams, Arndt Von Haeseler, and Robert Lanfear. 2020. ‘IQ-TREE 2: New Models and Efficient Methods for Phylogenetic Inference in the Genomic Era’. Edited by Emma Teeling. Molecular Biology and Evolution 37 (5): 1530–34. https://doi.org/10.1093/molbev/msaa015.
NCBI. 1988. ‘Nucleotide [Internet]. Bethesda (MD): National Library of Medicine (US), National Center for Biotechnology Information.’ https://www.ncbi.nlm.nih.gov/nucleotide/.
Paradis, Emmanuel, and Klaus Schliep. 2019. ‘Ape 5.0: An Environment for Modern Phylogenetics and Evolutionary Analyses in R’. Bioinformatics 35:526–28. https://doi.org/10.1093/bioinformatics/bty633.
Pentinsaari, Mikko, Heli Salmela, Marko Mutanen, and Tomas Roslin. 2016. ‘Molecular Evolution of a Widely-Adopted Taxonomic Marker (COI) across the Animal Tree of Life’. Scientific Reports 6 (1): 35275. https://doi.org/10.1038/srep35275.
Price, M. N., P. S. Dehal, and A. P. Arkin. 2009. ‘FastTree: Computing Large Minimum Evolution Trees with Profiles Instead of a Distance Matrix’. Molecular Biology and Evolution 26 (7): 1641–50. https://doi.org/10.1093/molbev/msp077.
R Core Team. 2022. R: A Language and Environment for Statistical Computing. Vienna, Austria: R Foundation for Statistical Computing. https://www.R-project.org/.
Yu, Guangchuang, David K. Smith, Huachen Zhu, Yi Guan, and Tommy Tsan‐Yuk Lam. 2017. ‘GGTREE : An R Package for Visualization and Annotation of Phylogenetic Trees with Their Covariates and Other Associated Data’. Edited by Greg McInerny. Methods in Ecology and Evolution 8 (1): 28–36. https://doi.org/10.1111/2041-210X.12628.


